### ISSUE TITLE
RHVA example notebook crashes. Exploding norm during leapfrog

### ISSUE THREAD

**Issue by Theaublanchard** [clementchadebec/benchmark_VAE#162]
**Bug description**
During training of the RHVAE on the simplest case, the sampling ultimately yields latent sample of exploding magnitude. Thus the evaluation of the log density of the sample returns NaN values.

More precisely, the error occurs because the method `_log_z` from the RHVAE class is called with infinite arguments. Those come from the second step of the leapfrog integrator (`_leap_step_2`).
Upon logging the norm of the gradient of the Hamiltonian, it explodes randomly after some time. The same phenomenon appears when using the "bce" reconstruction loss, albeit the training is stable for longer.

**To Reproduce**
1. Run the *example notebook* at [rhvae_training.ipynb](https://github.com/clementchadebec/benchmark_VAE/blob/main/examples/notebooks/models_training/rhvae_training.ipynb)
2. The training will fail after a few epochs (see error stack below). Decreasing the learning rate only delays the crash.

**Expected behavior**
No crash during training

**Desktop**
 - OS: SUSE Linux Enterprise Server 15 SP6
 - python: 3.12.4
 - numpy: 1.26.4
 - torch: 2.5.1

**Possible fix**
Clip the norm of the gradient on each call to the gradient of the Hamiltonian. It could disturb the training, this needs to be explored.

<img width="716" alt="Image" src="https://github.com/user-attachments/assets/5038edcb-4d53-4b9c-8bfc-0dd4ac9cd907" />

<img width="717" alt="Image" src="https://github.com/user-attachments/assets/a76a5dc6-efd9-48aa-9db6-29261f5e828c" />

<img width="885" alt="Image" src="https://github.com/user-attachments/assets/995a379e-4c0b-4ef1-a7c9-f7494182e8fe" />

![Image](https://github.com/user-attachments/assets/fdc57359-b417-407c-9263-be29dd03e349)
